### DISCUSSION TITLE
SELECT DISTINCT

### DISCUSSION THREAD

**Discussion by krzysztofRawski** [supabase/supabase#3294] [Questions] [answered]
How can I SELECT DISTINCT with Supabase API?

**Comment by dshukertjr** [comment, marked as answer]
Hi @krzysztofRawski !

You can create a view to achieve it!

**Reply by krzysztofRawski** [reply]
A view? What do you mean by that? Can you sand a link to a doc or something?

**Reply by dshukertjr** [reply]
Yeah, [here](https://supabase.io/blog/2020/11/18/postgresql-views) is a Supabase blog post explaining in detail what views are, but in short views are like shortcuts for a query. The nice thing about view is that with Supabase, you can query against views the same way you would against tables.

For example, let's say you have a table named `customer` that holds bunch of customer information. You can create a view to get distinct countries of every customers like this:

```sql
create view distinct_country as
    select distinct country from customers;
```

You can get the contents of the view like this 

```typescript
const { data, error } = await supabase.from('distinct_country').select();
```

And that's it! You can perform `distinnct` query with Supabase!

**Reply by krzysztofRawski** [reply]
Thank you very much.

**Reply by mellowcello77** [reply]
Views are so cool - a distinct could work in the front api's.

**Reply by rahul31agrawal** [reply]
How can we write this code in Python

**Reply by Agustin2214** [reply]
Is it possible to pass parameters to it?

**Reply by steve-chavez** [reply]
@Agustin2214 You can use a function for that. All filters work the same on functions (RPC) as on views.

**Comment by dotku** [comment]
> Views are so cool - a distinct could work in the front api's.

do you have sample code, I want to distinct based on one of the columns; how to achieve it?

**Comment by cvega21** [comment]
Am I missing something here? Why is such a basic SQL feature not included in the Supabase API?

**Reply by Klrfl** [reply]
I think sometime in the past the SDK used to be able to do this, with the options in the `.select()` method. Don't know why it got dropped though

https://www.restack.io/docs/supabase-knowledge-supabase-select-distinct

**Reply by Splutonics** [reply]
To create the view, looks like you still need to use SQL Editor. After that, it shows up under Table Editor in the UI. 
You reference a view the same way you reference a table- for example for at least the supabase javascript library, you would get the view like this.
`const { data: viewData } = await supabase.from("my_custom_view").select("*")`

**Reply by antonio-bluco** [reply]
@Splutonics but you cannot Join() views with other tables, which feels like a pretty major limitation. See my other comment below :(

**Comment by antonio-bluco** [comment]
But how to deal with the fact that views have no foreign keys? 

I want to write a view, get my distinct rows, and then join() with other tables. But views have no foreign keys, so this is not possible. So I have to do the join on the view directly, which has its own problems (issue with column names needing to be unique (leading to manual aliasing of all the column names, which gets very verbose very fast), flat structure instead of a nice JSON structured message that you would get from the supabase client, etc.). I have written more about this [here](https://www.reddit.com/r/Supabase/comments/1e8jmmx/how_to_work_with_views_since_they_have_no_foreign/). 

So how do you work with views in practice? This seems to be the standard suggestion to go around postgREST limitations, but they seem like a huge pain to work with.

**Reply by dshukertjr** [reply]
>So I have to do the join on the view directly

This is the suggested path. You can achieve flat structures using aggregate functions like `json_agg` or `json_object_agg`.
https://www.postgresql.org/docs/9.5/functions-aggregate.html
